Use Tera templates for rustdoc.

jsha opened the pull request:
Replaces a format!() call in layout::render with a template
expansion. Introduces a `templates` field in SharedContext so parts
of rustdoc can share pre-rendered templates.

This currently builds in a copy of the single template available, like
with static files. However, future work can make this live-loadable with
a perma-unstable flag, to make rustdoc developers' work easier.

Part of #84419.

Demo at https://hoffman-andrews.com/rust/tera/std/string/struct.String.html.

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-highfive commented:
r? @GuillaumeGomez

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2779671334) [(plain)](https://github.com/rust-lang/rust/commit/311c4388108b9a1484bd3f208fe3ef6a4b80e361/checks/2779671334/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling tempfile v3.2.0
   Compiling matchers v0.0.1
   Compiling globset v0.4.5
   Compiling ignore v0.4.17
   Compiling globwalk v0.8.1
   Compiling tracing-attributes v0.1.13
   Compiling pest_derive v2.1.0
   Compiling tracing v0.1.25
   Compiling tracing-subscriber v0.2.16
---
failures:

---- [rustdoc] rustdoc/static-root-path.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/static-root-path" "/checkout/src/test/rustdoc/static-root-path.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
13: @matches check failed
 `PATTERN` did not match
 // @matches - '"\.\./\.\./source-files.js"'
14: @!matches check failed
 `PATTERN` did not match
 // @!matches - '"/cache/source-files\.js"'
Encountered 2 errors

------------------------------------------

---
test result: FAILED. 441 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 24.10s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:12:08
```

</details>

jyn514 commented on Cargo.lock:
```diff
@@ -1482,6 +1482,17 @@ dependencies = [
 ]
 
+[[package]]
+name = "globwalk"
```
Oh oof I forgot tera uses globwalk :( it's extremely slow: https://github.com/rust-lang/docs.rs/pull/861/commits/9c9647ded669057493186245f5dcbc3b248d2783

jsha replied:
We don't actually use `globwalk` in this PR, we are manually loading files the same way the linked docs.rs PR does.

GuillaumeGomez commented:
For now, I don't see much advantages to this change, but at least it's a good start. What about the size of generated files and performance comparison?

jyn514 commented:
@GuillaumeGomez the advantage is it's much easier to read than repeated write calls

GuillaumeGomez commented:
In this case not really since only the big one has been moved.

jsha commented:
It's true this is mainly an initial setup to make subsequent templatizations possible, and to give a feel for what the template language will feel like. And the one format call here is the most template-like thing we currently have, since it already has named fields for interpolation. But even in this initial change you can see some of the benefits in clarity and readability. Consider:

Old:

```
        logo = {
            if layout.logo.is_empty() {
                format!(
                    "<a href='{root}{path}index.html'>\
                     <div class='logo-container rust-logo'>\
                     <img src='{static_root_path}rust-logo{suffix}.png' alt='logo'></div></a>",
                    root = page.root_path,
                    path = ensure_trailing_slash(&layout.krate),
                    static_root_path = static_root_path,
                    suffix = page.resource_suffix
                )
            } else {
                format!(
                    "<a href='{root}{path}index.html'>\
                     <div class='logo-container'><img src='{logo}' alt='logo'></div></a>",
                    root = page.root_path,
                    path = ensure_trailing_slash(&layout.krate),
                    logo = layout.logo
                )
            }
        },
```

New:

```
<a href='{{page.root_path | safe}}{{krate_with_trailing_slash | safe}}index.html'>
    <div class='logo-container'>
    <img src='
        {%- if layout.logo -%}
        {{layout.logo}}
        {%- else -%}
        {{static_root_path | safe}}rust-logo{{page.resource_suffix}}.png
        {%- endif -%}
        ' alt='logo'>
    </div>
</a>
```

That's 20 lines vs 11 lines, and the 11 line version also removes some redundancy. Also, the HTML in the template version is easier to read without backslashes in front of all the quotes.

Performance: The difference so far will be quite negligible, because the amount of code moved in this PR is small. As we discussed in #84419, we may need to take a bit of a leap of faith performance-wise since the overall performance of templatizing everything won't be clear until we're much further along.

Byte size: struct.String.html is 54,713 bytes gzipped, vs 54,594 on master. Note that this branch doesn't have special treatment for removing excess whitespace. I used `{%-` and `-%}` where appropriate to take advantage of newline control, but the natural layout includes newlines and indents even between constant HTML elements, where there's no natural place for `{%-` and `-%}`. I think as a followup we should plan to pass the templatized content through a very basic parser that can remove HTML whitespace. If you'd like that as a prereq to landing this branch, let me know.

GuillaumeGomez commented:
> Byte size: struct.String.html is 54,713 bytes gzipped, vs 54,594 on master. Note that this branch doesn't have special treatment for removing excess whitespace. I used `{%-` and `-%}` where appropriate to take advantage of newline control, but the natural layout includes newlines and indents even between constant HTML elements, where there's no natural place for `{%-` and `-%}`. I think as a followup we should plan to pass the templatized content through a very basic parser that can remove HTML whitespace. If you'd like that as a prereq to landing this branch, let me know.

Yes please. I'm honestly very afraid of the whole "remove whitespaces from HTML". In this case, we even need to re-parse it. Even though I really like the whole templating idea, the downside is really worrying me. But I may be completely wrong. So I'd prefer to have the whitespaces removed from HTML before merging.

jyn514 commented:
@GuillaumeGomez are you worried about reparsing? or about the change in file size? If you're just worried about reparsing, I don't see why this has to be blocked on removing the whitespace.

jsha commented:
Alright, with the latest updated, `struct.String.html` is 54,597 gzipped bytes, 2 bytes smaller than on master. :-)

No parsing required! The trick was to add `{#- -#}` to the end of each line. `{#` is a comment, and adding `-` triggers the whitespace removal path.

This is still not completely optimal, since it incentivizes putting a whole HTML tag on a single line, even when you might prefer to split out attributes on their own line. If you split out attributes on their own line, you're forced to choose between putting `{#- -#}` at the end of each line, which will concatenate the attributes and horribly break everything, or not putting any whitespace control in at all and incurring some extra bytes. Still, I think it's an okay tradeoff.

One thing that might be cool longer term is to see about getting [Django's spaceless](https://docs.djangoproject.com/en/3.2/ref/templates/builtins/#spaceless) feature implemented in Tera (though I think we'd also want a special mode that's aware of HTML templates and collapses N whitespaces into 1 space character).

jyn514 commented:
> One thing that might be cool longer term is to see about getting Django's spaceless feature implemented in Tera (though I think we'd also want a special mode that's aware of HTML templates and collapses N whitespaces into 1 space character).

That would be amazing! Can you open an issue on tera?

jsha commented:
Done: https://github.com/Keats/tera/issues/641

I also figured out how to do single whitespace for attributes: `{# -#}` instead of `{#- -#}` (that is, remove whitespace to the right, but not to the left). Switched to that, and added a STYLE.md discussing style for templates.

GuillaumeGomez commented on src/librustdoc/html/templates/page.html:
```diff
@@ -0,0 +1,119 @@
+    <meta name="keywords" content="{{page.keywords}}"> {#- -#}
+    <title>{{page.title}}</title> {#- -#}
+    <link rel="stylesheet" type="text/css" {# -#}
+          href="{{static_root_path | safe}}normalize{{page.resource_suffix}}.css"> {#- -#}
```
This is weird to have two different syntaxes depending if we're in a tag or not. Oh well...

bors commented:
:hourglass: Testing commit cd0f93193c84ddc6698f9b65909da71c084dcb74 with merge 9d93819fa7ae40376ac1f563a858d2ba5f3b0277...
<!-- homu: {"type":"BuildStarted","head_sha":"cd0f93193c84ddc6698f9b65909da71c084dcb74","merge_sha":"9d93819fa7ae40376ac1f563a858d2ba5f3b0277"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2875051400)
Approved by: jyn514,GuillaumeGomez
Pushing 9d93819fa7ae40376ac1f563a858d2ba5f3b0277 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514,GuillaumeGomez","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2875051400"},"merge_sha":"9d93819fa7ae40376ac1f563a858d2ba5f3b0277"} -->
